rad/prelude/patterns.rad

```diff
@@ -0,0 +1,192 @@
+;; Because pattern matching uses the Maybe-monad, it's helpful to define some
+;; essential Maybe-monad function now.
+
+(def maybe->>=
+  "Monadic bind for the maybe monad."
+  (fn [v f]
+    (if (vector? v)
+      (if (eq? (length v) 2)
+        (if (eq? (nth 0 v) :Just)
+          (f (nth 1 v))
+          :Nothing)
+        :Nothing)
+      :Nothing)))
+
+(def-rec maybe-foldlM
+  "Monadic fold over the elements of a seq, associating to the left (i.e. from
+  left to right) in the maybe monad."
+  (fn [f i xs]
+    (if (empty-seq? xs)
+      [:Just i]
+      (maybe->>= (f i (nth 0 xs))
+                 (fn [b] (maybe-foldlM f b (drop 1 xs)))))))
+
+(def exclusive-dict-merge
+  "Merges two dicts while checking for key conflicts. Returns `{:merge
+  m :conflicts c}` where `m` is the merged dict for all non-conflicting keys and
+  `c` is a dict with all the conflicting keys, mapping to pairs of values, one
+  from each input dict."
+  (fn [m n]
+    (foldr (fn [kv acc]
+             (def k (nth 0 kv))
+             (def v (nth 1 kv))
+             (def merge (lookup :merge acc))
+             (if (member? k merge)
+               {:merge     (delete k merge)
+                :conflicts (insert k [(lookup k merge) v] (lookup :conflicts acc))}
+               (insert :merge (insert k v merge) acc)))
+           {:merge m
+            :conflicts {}}
+           (seq n))))
+
+(def non-linear-merge
+  "The bindings merge strategy for non-linear patterns. Use this function to merge
+  bindings returned by sub-patterns if you want your pattern to be non-linear."
+  (fn [m n]
+    (def excl-merge (exclusive-dict-merge m n))
+    (def cs (lookup :conflicts excl-merge))
+    (def merge (lookup :merge excl-merge))
+    (if (eq? cs {})
+      [:Just merge]
+      (maybe-foldlM (fn [acc kv]
+                      (def v (nth 1 kv))
+                      (def v0 (nth 0 v))
+                      (if (eq? v0 (nth 1 v))
+                        [:Just (insert (nth 0 kv) v0 acc)]
+                        :Nothing))
+                    merge
+                    (seq cs)))))
+(def _
```

Comment:
Hm, do we need this? If `_` remains a normal identifier, don't we get exactly the same behavior (modulo `_` actually being bound)?

Reply:
Oh yeah. Then can we maybe name it `\_`? That'd make it clear it's pattern-specific, and leave `_` for other uses.

Reply:
Hm, actually, that might be quite a bit uglier, nvm.